src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,778 @@
+                    0.4928777813127657, 0.4267340097946024, 0.36914706729636887, 0.3190553736355825, 0.27552277516026125, 0.23772456873189068,
+                    0.20493497304473338, 0.17651591132190647, 0.1519069804835684, 0.13061649224726435, 0.11221348131208278, 0.09632058481723846,
+                    0.08260770567516164, 0.0707863801843716, 0.06060477755511267, 0.051843265658779024, 0.0443104834690419, 0.03783986632710667,
+                    0.03228657536442549, 0.027524787181948417, 0.02344530424356765, 0.019953450420057577, 0.01696721974494692, 0.014415649740821513,
+                    0.012237393667929978, 0.010379468759906684, 0.008796159966022614, 0.0074480609365136455, 0.006301235986898177,
+                    0.00532648857725966, 0.004498723460523362, 0.0037963911059268884, 0.0032010043051660104, 0.002696718032995797,
+                    0.0022699646742388863, 0.0019091376570554135, 0.0011570531254881296, 0.000697019955113331, 0.00041737721863073713,
+                    0.000248438820613534, 0.00014700521929794912, 8.647365841055832e-05, 5.056939088336744e-05, 2.9400808653120604e-05,
+                    1.6994687082728674e-05, 9.767061541798089e-06
+                };
+
+            private readonly double _threshold;
+            private readonly double _sensitivity;
+            private readonly SrCnnDetectMode _detectMode;
+            private readonly int _outputLength;
+
+            public SrCnnEntireModeler(double threshold, double sensitivity, SrCnnDetectMode detectMode, int outputLength)
+            {
+                _threshold = threshold;
+                _sensitivity = sensitivity;
+                _detectMode = detectMode;
+                _outputLength = outputLength;
+            }
+
+            public double[][] Train(double[] values)
+            {
+                double[][] results = new double[values.Length][];
+                for (int i = 0; i < results.Length; ++i)
+                {
+                    results[i] = new double[_outputLength];
+                }
+                SpecturalResidual(values, results, _threshold);
+                //Optional Steps
+                if (_detectMode == SrCnnDetectMode.AnomalyAndMargin)
+                {
+                    GetMargin(values, results, _sensitivity);
+                }
+                else if (_detectMode == SrCnnDetectMode.AnomalyAndExpectedValue)
+                {
+                    GetExpectedValue(values, results);
+                }
+                return results;
+            }
+
+            private static void SpecturalResidual(double[] values, double[][] results, double threshold)
+            {
+                // Step 1: Get backadd wave
+                double[] backAddList = BackAdd(values);
+
+                // Step 2: FFT transformation
+                int length = backAddList.Length;
+                double[] fftRe = new double[length];
```

Comment:
>fftRe [](start = 25, length = 5)

Is there a way to make these fields of the class so they are only allocated once?

Reply:
I would make this class contain buffers for all these computations to avoid multiple allocations of arrays. This would mean that the `SrCnnBatchAnomalyDetector` cannot have an `SrcEntireModeler` as a field, and the `InitializeBatch` method would have to instantiate it (`IDataVew`s should be thread safe, since it should be possible to create multiple row cursors on multiple threads from an `IDataView` object).

---
In reply to: [426420411](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426420411) [](ancestors = 426420411)

Reply:
I think the problem here is the same as results array, due to the batchSizxe=-1 scenario, we will not know the length of these arrays until we actually read in data.

---
In reply to: [426431891](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426431891) [](ancestors = 426431891,426420411)

Reply:
Allocating a bunch of arrays for each batch might have a perf price, with the GC going into overdrive. Consider allocating them only once and re-using them for subsequent batches.

---
In reply to: [427139611](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427139611) [](ancestors = 427139611,426431891,426420411)

Reply:
If every instance of `SrEntireModeler` contained preallocated buffers, would the result not be the same (since `InitializeBatch` would then have to allocate a new instance of `SrEntireModeler`)?

Reply:
No, because `InitializeBatch` is called only once, when the row cursor is created (see line 100 in BatchDataViewMapperBase.cs). The `SrEntireModeler` methods (`Train` and the helper methods that are called from it) are called after every batch is read from the input cursor (see line 168 in BatchDataViewMapperBase.cs).

---
In reply to: [427614415](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427614415) [](ancestors = 427614415)

Reply:
Perhaps `InitializeBatch` should be renamed `CreateBatch` to avoid confusion?

---
In reply to: [427764070](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427764070) [](ancestors = 427764070,427614415)

Reply:
done.

---
In reply to: [427764306](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427764306) [](ancestors = 427764306,427764070,427614415)

Reply:
fixed.

---
In reply to: [427285088](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427285088) [](ancestors = 427285088,427139611,426431891,426420411)